File: src/linkage_sim/solvers/assembly.py
"""Global constraint system assembly from a Mechanism.

Assembles the global Φ, Φ_q, and γ vectors/matrices by stacking
contributions from all joints in the mechanism.
"""

from __future__ import annotations

import numpy as np
from numpy.typing import NDArray

from linkage_sim.core.mechanism import Mechanism


def assemble_constraints(
    mechanism: Mechanism, q: NDArray[np.float64], t: float
) -> NDArray[np.float64]:
    """Assemble global constraint residual vector Φ(q, t).

    Returns:
        Φ: (m,) vector where m = total constraint equations.
    """
    m = mechanism.n_constraints
    phi = np.zeros(m)

    row = 0
    for joint in mechanism.joints:
        n_eq = joint.n_equations
        phi[row : row + n_eq] = joint.constraint(mechanism.state, q, t)
        row += n_eq

    return phi


def assemble_jacobian(
    mechanism: Mechanism, q: NDArray[np.float64], t: float
) -> NDArray[np.float64]:
    """Assemble global constraint Jacobian Φ_q(q, t).

    Returns:
        Φ_q: (m, n_coords) matrix.
    """
    m = mechanism.n_constraints
    n = mechanism.state.n_coords
    phi_q = np.zeros((m, n))

    row = 0
    for joint in mechanism.joints:
        n_eq = joint.n_equations
        phi_q[row : row + n_eq, :] = joint.jacobian(mechanism.state, q, t)
        row += n_eq

    return phi_q


def assemble_gamma(
    mechanism: Mechanism,
    q: NDArray[np.float64],
    q_dot: NDArray[np.float64],
    t: float,
) -> NDArray[np.float64]:
    """Assemble global acceleration RHS vector γ(q, q̇, t).

    Used in: Φ_q * q̈ = γ

    Returns:
        γ: (m,) vector.
    """
    m = mechanism.n_constraints
    gamma = np.zeros(m)

    row = 0
    for joint in mechanism.joints:
        n_eq = joint.n_equations
        gamma[row : row + n_eq] = joint.gamma(mechanism.state, q, q_dot, t)
        row += n_eq

    return gamma

File: linkage-sim-rs/src/solver/assembly.rs
//! Global constraint system assembly from a Mechanism.
//!
//! Assembles the global Φ, Φ_q, Φ_t, and γ vectors/matrices by stacking
//! contributions from all joints in the mechanism.

use nalgebra::{DMatrix, DVector};

use crate::core::mechanism::Mechanism;

/// Assemble global constraint residual vector Φ(q, t).
pub fn assemble_constraints(mech: &Mechanism, q: &DVector<f64>, t: f64) -> DVector<f64> {
    let m = mech.n_constraints();
    let mut phi = DVector::zeros(m);
    let state = mech.state();

    let mut row = 0;
    for joint in mech.all_constraints() {
        let n_eq = joint.n_equations();
        let c = joint.constraint(state, q, t);
        for i in 0..n_eq {
            phi[row + i] = c[i];
        }
        row += n_eq;
    }

    phi
}

/// Assemble global constraint Jacobian Φ_q(q, t).
pub fn assemble_jacobian(mech: &Mechanism, q: &DVector<f64>, t: f64) -> DMatrix<f64> {
    let m = mech.n_constraints();
    let n = mech.state().n_coords();
    let mut phi_q = DMatrix::zeros(m, n);
    let state = mech.state();

    let mut row = 0;
    for joint in mech.all_constraints() {
        let n_eq = joint.n_equations();
        let jac = joint.jacobian(state, q, t);
        for i in 0..n_eq {
            for j in 0..n {
                phi_q[(row + i, j)] = jac[(i, j)];
            }
        }
        row += n_eq;
    }

    phi_q
}

/// Assemble global time-derivative vector Φ_t(q, t).
pub fn assemble_phi_t(mech: &Mechanism, q: &DVector<f64>, t: f64) -> DVector<f64> {
    let m = mech.n_constraints();
    let mut phi_t = DVector::zeros(m);
    let state = mech.state();

    let mut row = 0;
    for joint in mech.all_constraints() {
        let n_eq = joint.n_equations();
        let pt = joint.phi_t(state, q, t);
        for i in 0..n_eq {
            phi_t[row + i] = pt[i];
        }
        row += n_eq;
    }

    phi_t
}

/// Assemble global acceleration RHS vector γ(q, q̇, t).
pub fn assemble_gamma(
    mech: &Mechanism,
    q: &DVector<f64>,
    q_dot: &DVector<f64>,
    t: f64,
) -> DVector<f64> {
    let m = mech.n_constraints();
    let mut gamma = DVector::zeros(m);
    let state = mech.state();

    let mut row = 0;
    for joint in mech.all_constraints() {
        let n_eq = joint.n_equations();
        let g = joint.gamma(state, q, q_dot, t);
        for i in 0..n_eq {
            gamma[row + i] = g[i];
        }
        row += n_eq;
    }

    gamma
}
